src/main/java/org/folio/services/consortia/ShadowInstanceSynchronizationHandler.java

```diff
@@ -0,0 +1,168 @@
+      instanceId, headers)
+      .compose(tenantIds -> updateShadowInstances(event, tenantIds, headers));
+  }
+
+  private Future<List<String>> getShadowInstancesTenantIds(String consortiumId, String centralTenantId,
+                                                           String instanceId, Map<String, String> headers) {
+    String okapiUrl = headers.get(URL);
+    String preparedPath = format(SHARING_INSTANCES_PATH, consortiumId, instanceId);
+    WebClient client = WebClient.wrap(httpClient);
+    HttpRequest<Buffer> request = client.requestAbs(GET, okapiUrl + preparedPath);
+    headers.forEach(request::putHeader);
+    request.addQueryParam(LIMIT_QUERY_PARAM, TENANT_IDS_LIMIT);
+
+    return request.send().compose(response -> {
+      if (response.statusCode() != HTTP_OK) {
+        String msg = format("Error retrieving shadow instances tenant ids, response status: '%s', response body: '%s'",
+          response.statusCode(), response.bodyAsString());
+        LOG.warn("getShadowInstancesTenantIds:: {}", msg);
+        return Future.failedFuture(msg);
+      }
+
+      List<String> affiliationsTenantIds = response.bodyAsJsonObject().getJsonArray(SHARING_INSTANCES_FIELD)
+        .stream()
+        .map(JsonObject.class::cast)
+        .flatMap(sharing ->
+          Stream.of(sharing.getString(TARGET_TENANT_ID_FIELD), sharing.getString(SOURCE_TENANT_ID_FIELD)))
+        .filter(tenantId -> !tenantId.equals(centralTenantId))
+        .collect(Collectors.toList());
+      return Future.succeededFuture(affiliationsTenantIds);
+    });
+  }
+
+  private Future<Void> updateShadowInstances(DomainEvent<Instance> event, List<String> tenantIds,
+                                             Map<String, String> headers) {
+    LOG.info("updateShadowInstances:: Trying to update shadow instances in the following tenants: {} ", tenantIds);
+    ArrayList<Future<Void>> updateFutures = new ArrayList<>();
+    Instance instance = JsonObject.mapFrom(event.getNewEntity()).mapTo(Instance.class);
+    prepareInstanceForUpdate(instance);
+
+    for (String tenantId : tenantIds) {
+      updateFutures.add(updateShadowInstance(instance, tenantId, headers));
+    }
+    return GenericCompositeFuture.join(updateFutures).mapEmpty();
+  }
+
+  private void prepareInstanceForUpdate(Instance instance) {
+    instance.setSource(format(CONSORTIUM_SOURCE_TEMPLATE, instance.getSource()));
+  }
+
+  private Future<Void> updateShadowInstance(Instance instance, String tenantId, Map<String, String> okapiHeaders) {
+    LOG.debug("updateShadowInstance:: Trying to update shadow instance, tenantId: '{}', instanceId: '{}'",
+      tenantId, instance.getId());
+    HashMap<String, String> headers = new HashMap<>(okapiHeaders);
+    headers.put(TENANT, tenantId);
+    InstanceRepository instanceRepository = new InstanceRepository(vertx.getOrCreateContext(), headers);
+
+    return instanceRepository.getById(instance.getId())
```

Comment:
Can we have retries when the update is blocked by optimistic locking? or do you think that is unnecessary and is not a pattern used in other parts of inventory?

Reply:
it might be unnecessary because updating shadow instances through the UI or data import is not expected. There is a probability of encountering optimistic locking error during a shared instance update using ui, but in this case, the shared instance update will fail and this handler will not be called.

Reply:
Sounds good!